### ISSUE TITLE
How to port to a new CPU platform?

### ISSUE THREAD

**Issue by snikeguo** [dotnet/runtime#47256]
How to port to a new CPU platform?
I want to port coreclr to MCU platform, such as  NXP I.mx.rt  series /  ST  Stm32h7 series (arm cortex-m7, 600 + MHz, thumb2 instruction , chip contains 1MB SRAM, 4MB NOR Flash), MCU can access external SDRAM, nor flash. and it can run programs directly in NOR flash. (by the way, 64MB SDRAM is about $0.7.)
So I am reading the PAL source code of coreclr, and I have encountered many problems that API functions cannot be transplanted, such as:
```
1.PALAPI CreateThread(
IN LPSECURITY_ ATTRIBUTES lpThreadAttributes,
IN DWORD dwStackSize,
IN LPTHREAD_ START_ ROUTINE lpStartAddress,
IN LPVOID lpParameter,
IN DWORD dwCreationFlags,
OUT LPDWORD lpThreadId)；
```
I think you should have forked the CLR code of. Net framework to make the dotnet project development faster. Similar to these “  LPSECURITY_ ATTRIBUTES lpThreadAttributes” parameters bring great difficulties to porting. For FreeRTOS, the creation thread has the following parameters:
```
BaseType_ t xTaskCreate(
TaskFunction_ tp vTaskCode,//entry
const char * constpcName,//thread name
unsigned short usStackDepth,//stack size
void *pvParameters,//thread arg
UBaseType_ t uxPriority,//priority
TaskHandle_ t *pvCreatedTask
)；
```
2.
```
PALAPI
MessageBoxW(
IN LPVOID hWnd,  // NOTE: diff from winuser.h
IN LPCWSTR lpText,
IN LPCWSTR lpCaption,
IN UINT uType);
```
Because there is no GUI
### I think you should make a migration Manual of minimum coreclr to support JIT, GC and thread.
3. Why don't I choose nanoframework?
Because it's too slow. Nanoframework is interpreted and executed without JIT. But SDRAM is big enough, MCU can use JIT completely.
4.ReadyToRun(R2R) For MCU?
input： ilcode、HelloWorld Image Offset
out put:(HelloWorld NativeIamge+ILCode).hex
```
HelloWorld IL code--------------------> |         |	
                                        |  R2R.exe | ------->(HelloWorld NativeIamge+ILCode).hex
HelloWorld Image Offset --------------->|          |
```
then download to mcu:
```
nor flash:
  8000000 -------------------------------80010000---------
  NativeDriver+CoreClr   |	align 	 |   HelloWorld NativeIamge+ILCode
  ```
In this way, MCU directly runs HelloWorld native image in NOR flash

**Comment by hoyosjs** [comment]
@davidwrighton helped with writing a rough overview of what needs to be done to bring support https://github.com/dotnet/runtime/blob/master/docs/design/coreclr/botr/guide-for-porting.md We know to some extent the PAL was more of a "make system API look like windows" more than "here is a base functionality needed to support different OS's", and there's been conversations of what can be done around that. For example, what used to be CoreRT did more of a layer of functionality. The PAL in coreclr comes from a lot of history and different projects - and while not ideal and a bit organic in that sense - it works pretty well for most cases. A lot of these emulations can be made work pretty well, as seen with all the Linux distros that have been supported under it. As for Native AOT as you point out, it's something that has been investigated and continues to be - we consider it part of the form factors as described  in [.NET Runtime Form Factors](https://github.com/dotnet/designs/blob/main/accepted/2020/form-factors.md#ahead-of-time-aot-compilation), and there's still some active work in ways to provide better experiences in that area. Maybe @janvorli also has some tips for the PAL layer for such an OS?

**Comment by snikeguo** [comment]
https://docs.microsoft.com/en-us/cpp/build/overview-of-arm-abi-conventions?view=msvc-160
‘ The instruction set for Windows on ARM is strictly limited to Thumb-2 ’
It seems that coreclr has completed the JIT compiler of thumb-2 instruction set?
If so, does it mean that I only need to migrate pal layer code?


Can the coreclr team optimize the PAL layer code architecture to make coreclr more portable?
such as:
```
HANDLE
PALAPI
CreateFileW(
        IN LPCWSTR lpFileName,
        IN DWORD dwDesiredAccess,
        IN DWORD dwShareMode,
        IN LPSECURITY_ATTRIBUTES lpSecurityAttributes,
        IN DWORD dwCreationDisposition,
        IN DWORD dwFlagsAndAttributes,
        IN HANDLE hTemplateFile);
```
Some projects have no file system. only RTOS（only thread, queue, semaphore, mutex, task notification）
So I think PAL layer code architecture needs ：1. highly modular,2. portable 3.configurable .

**Comment by snikeguo** [comment]
like this（the code comes from my mcu RPC framework architecture code）：
```
EmbedXrpc_Thread_t EmbedXrpc_CreateThread(const char* threadName, uint8_t priority, void (*Thread)(void*), void* Arg) ;
EmbedXrpc_Mutex_t EmbedXrpc_CreateSemaphore(const char* semaphoreName);
EmbedXrpc_Mutex_t EmbedXrpc_CreateMutex(const char* mutexName) ;
EmbedXrpc_Queue_t EmbedXrpc_CreateQueue(const char* queueName, uint32_t queueItemSize, uint32_t maxItemLen) ;
EmbedXrpc_Timer_t EmbedXrpc_CreateTimer(const char* timerName, uint32_t timeout, void (*timercb)(void* arg), void* Arg) ;

void EmbedXrpc_DeleteThread(EmbedXrpc_Thread_t thread) ;
void EmbedXrpc_DeleteMutex(EmbedXrpc_Mutex_t) ;
void EmbedXrpc_DeleteQueue(EmbedXrpc_Queue_t) ;
void EmbedXrpc_DeleteSemaphore(EmbedXrpc_Semaphore_t) ;
void EmbedXrpc_DeleteTimer(EmbedXrpc_Timer_t) ;

void EmbedXrpc_ThreadStart(EmbedXrpc_Thread_t thread) ;

void EmbedXrpc_TimerStart(EmbedXrpc_Timer_t timer, uint16_t interval) ;
void EmbedXrpc_TimerReset(EmbedXrpc_Timer_t timer) ;
void EmbedXrpc_TimerStop(EmbedXrpc_Timer_t timer) ;

bool EmbedXrpc_TakeSemaphore(EmbedXrpc_Semaphore_t sem, uint32_t timeout) ;
void EmbedXrpc_ReleaseSemaphore(EmbedXrpc_Semaphore_t sem) ;
void EmbedXrpc_ResetSemaphore(EmbedXrpc_Semaphore_t sem) ;

bool EmbedXrpc_TakeMutex(EmbedXrpc_Mutex_t mutex, uint32_t timeout) ;
bool EmbedXrpc_ReleaseMutex(EmbedXrpc_Mutex_t mutex) ;



QueueState EmbedXrpc_ReceiveQueue(EmbedXrpc_Queue_t queue, void* item, uint32_t itemSize, uint32_t timeout) ;
QueueState EmbedXrpc_SendQueue(EmbedXrpc_Queue_t queue, void* item, uint32_t itemSize) ;
void EmbedXrpc_ResetQueue(EmbedXrpc_Queue_t queue) ;
uint32_t EmbedXrpc_QueueSpacesAvailable(EmbedXrpc_Queue_t queue);
void* Malloc(uint32_t size) ;
void Free(void* ptr) ;
void Memcpy(void* d, const void* s, uint32_t size) ;
```

**Comment by hoyosjs** [comment]
Yes, CoreCLR is thumb aware. A lot of the ARMv7 sfp support is mostly best try, and might have some hidden bugs. It's unlikely the PAL will change any time soon. That would be a pretty large undertaking. The PAL would still have file api's, other with things like System.File.IO would make no sense. What you are asking is a bit more extensive - basically for such OS's a lot  of API's in the .NET surface would need to be marked as disabled, given the platform wouldn't support them (or support them by creating an abstraction on top of memory or something). That's quite more than just a port.

**Comment by snikeguo** [comment]
@hoyosjs  OK, I'm looking forward to the new pal framework, otherwise I can only port the JVM to mcu, although I hate Java.
As for file operation, I think like rust language,the CORE only includes basic things, such as thread, collections, diagnostics, reflection... And STD includes file operation API.
In order to distinguish the API of core and STD, you can change the namespace of thread to Core.Threading.Thread , while the file operation is still in progress Std.Io (or System.Io )
hahaha.I don't think Microsoft will do that...........

**Comment by snikeguo** [comment]
I'm going to port the JVM to arm mcu.so close  the issue.

**Comment by hoyosjs** [comment]
Just to be clear, this PAL work is not planned for anytime soon as I am aware. And the "core"part that you say is a major rework of a lot of components. While it's great to hear there's interest for coreclr to be used in these scenarios, the team won't get to work on this in the foreseeable future in the current order of priorities. I was just trying to explain what it would entail to try to port to such architecture.

**Comment by snikeguo** [comment]
@hoyosjs  thank you very much for your answer

**Comment by snikeguo** [comment]
I'm going to try to read PAL code... and try to port it.

**Comment by snikeguo** [comment]
I'm going to try to read PAL code... and try to port it.
